### PR TITLE
[com_content] - add articles ordering by votes, ratings

### PR DESCRIPTION
**Pull request by alikon**
#### Summary of Changes

enable article ordering by votes or by rating on backend & frontend when content vote plugin is enabled
- [ ] added articles order options (vote asc, voted desc, rating asc, rating desc)

votes0
- [ ] added rating and vote columns

votes1
- [ ] added articles  Lists Layout options (show vote, show ratings)

vote2
- [ ] added articles order option in shared ( vote, rating)

votes3
- [ ] added column vote , rating on frontend list layout

votes4
- [ ] module articles category - added option to sort article by vote, rating 

votes5

history see #10989

### CODE REVIEW

**Comment by conconnl** [comment]
@alikon is this One ready to be retested by me and @brianteeman <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

File: administrator/components/com_content/models/articles.php [hunk 1]
```diff
@@ -313,23 +324,12 @@ protected function getListQuery()
...
-
-		if ($orderCol == 'access_level')
+		if (JPluginHelper::isEnabled('content', 'vote'))
 		{
-			$orderCol = 'ag.title';
+			$orderCol = empty($this->state->get('list.fullordering', 'a.id')) ? 'a.id' : $this->state->get('list.fullordering', 'a.id');
 		}
 
-		$query->order($db->escape($orderCol . ' ' . $orderDirn));
+		$query->order($db->escape($orderCol));
```
Comment (csthomas) [inline administrator/components/com_content/models/articles.php]: Should not we use $orderDirn?

File: administrator/components/com_content/models/articles.php [hunk 2]
```diff
@@ -199,13 +199,24 @@ protected function getListQuery()
 		$query->select('ua.name AS author_name')
 			->join('LEFT', '#__users AS ua ON ua.id = a.created_by');
 
+		// Join on voting table
+		$assogroup = 'a.id, l.title, l.image, uc.name, ag.title, c.title, ua.name';
+
+		if (JPluginHelper::isEnabled('content', 'vote'))
+		{
+			$assogroup = 'a.id, l.title, l.image, uc.name, ag.title, c.title, ua.name, v.rating_sum, v.rating_count';
```
Comment (csthomas) [inline administrator/components/com_content/models/articles.php]: `207`. What about add only additional columns?
```php
$assogroup .= ', v.rating_sum, v.rating_count';
```

Comment (alikon) [inline administrator/components/com_content/models/articles.php]: better

File: administrator/components/com_content/models/articles.php (hunk 1 above)
Comment (alikon) [inline administrator/components/com_content/models/articles.php]: thanks 
fixed

File: administrator/components/com_content/models/articles.php [hunk 3]
```diff
@@ -199,13 +199,24 @@ protected function getListQuery()
 		$query->select('ua.name AS author_name')
 			->join('LEFT', '#__users AS ua ON ua.id = a.created_by');
 
+		// Join on voting table
+		$assogroup = 'a.id, l.title, l.image, uc.name, ag.title, c.title, ua.name';
+
+		if (JPluginHelper::isEnabled('content', 'vote'))
```
Comment (csthomas) [inline administrator/components/com_content/models/articles.php]: IMHO I don't know you can use `JPluginHelper::isEnabled' here. May be better would be to use getState() and set it before that method.

File: administrator/components/com_content/models/articles.php [hunk 4]
```diff
@@ -199,13 +199,24 @@ protected function getListQuery()
...
+		{
+			$assogroup .= ', v.rating_sum, v.rating_count';
+			$query->select('COALESCE(NULLIF(ROUND(v.rating_sum  / v.rating_count, 0), 0), 0) AS rating, 
+					COALESCE(NULLIF(v.rating_count, 0), 0) as rating_count')
```
Comment (csthomas) [inline administrator/components/com_content/models/articles.php]: Does NULLIF is required?

File: administrator/components/com_content/models/articles.php (hunk 3 above)
Comment (alikon) [inline administrator/components/com_content/models/articles.php]: not sure
for  my POV  is more readable

File: administrator/components/com_content/models/articles.php (hunk 4 above)
Comment (alikon) [inline administrator/components/com_content/models/articles.php]: yes

**Comment by csthomas** [comment]
> JPluginHelper::isEnabled

What about using `class ContentModelArticles` in others modules / extensions? 
Do we have always load votes if the plugin is enabled?

**Comment by alikon** [comment]
@conconnl yes it is

**Comment by brianteeman** [comment]
Retested and found a few small issues

1. Please reorder the dropdown select so that it matches the column order ie move ID to the end
screen shot 2016-07-28 at 07 33 33

2. The visual indicator of the sort direction is being displayed on the ID column and not the rating or votes column
screen shot 2016-07-28 at 07 37 12<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 4dca61f7022cdcdd9cf71bb877462dbf0fa6d289<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

**Comment by csthomas** [comment]
For greater flexibility I suggest to add something like:
```php
		if ($params->get('list_show_votes', 0) || $params->get('list_show_ratings', 0))
		{
			$this->setState('load_votes', JPluginHelper::isEnabled('content', 'vote'));
		}
```
at populateState https://github.com/joomla/joomla-cms/pull/11225/files#diff-e5b3effc5f60b387662db42517e31654L129

This way modules won't need to load votes, better performance.
IMHO If some category does not like to show votes then sql for votes should not be added.

and at getListQuery method then you can use `if ($this->getState('load_votes', false))`.

**Comment by alikon** [comment]
@csthomas my branch is public feel free to submit a PR
@brianteeman fixed your reported issues

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on fc073ff463ab82ec8914239300193b926a469842<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

**Comment by conconnl** [comment]
I have tested this item :white_check_mark: successfully on fc073ff463ab82ec8914239300193b926a469842<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11225">issues.joomla.org/joomla-cms/11225</a>.</sub>

File: libraries/joomla/form/fields/list.php [hunk 5]
```diff
@@ -117,6 +117,12 @@ protected function getOptions()
 				{
 					continue;
 				}
+
+				// Requires vote plugin enabled
+				if (in_array('vote', $requires) && !JPluginHelper::isEnabled('content', 'vote'))
+				{
+					continue;
+				}
```
Comment (csthomas) [inline libraries/joomla/form/fields/list.php]: IMHO this is not a good place to do that.  

I do not know whether you can create a subclass of list and then use it, but it should be discuss.

You can not favor the core plugin, not core plugin can not do this.

**Comment by wilsonge** [comment]
Are these database queries compatible with postgres and mssql?

**Comment by alikon** [comment]
@wilsonge if your concern  are about `COALESCE()`  and  `NULLIF()` :

- for postgresql my answer is yes
- for mssql i'm unable to test but looking at msdn documentation  (yes)

https://msdn.microsoft.com/it-it/library/ms190349.aspx
https://msdn.microsoft.com/it-it/library/ms177562.aspx

**Comment by infograf768** [comment]
This gives a fatal error:
https://github.com/joomla/joomla-cms/issues/12095

**Comment by sanderpotjer** [comment]
@alikon while testing an other pull request I noticed that the new columns in the article manager haven't been implemented in the Hathor backend template. Is that on purpose?

**Comment by alikon** [comment]
I hoped that hathor would be removed from 3.7

**Comment by wilsonge** [comment]
No chance. Off you go do a PR :P

**Pull request merged**